src/include/storage/storage_structure/overflow_file.h

```diff
@@ -0,0 +1,159 @@
+#pragma once
+
+#include <fcntl.h>
+
+#include <cstdint>
+#include <memory>
+#include <string_view>
+#include <vector>
+
+#include "common/cast.h"
+#include "common/constants.h"
+#include "common/types/types.h"
+#include "storage/buffer_manager/bm_file_handle.h"
+#include "storage/buffer_manager/buffer_manager.h"
+#include "storage/file_handle.h"
+#include "storage/index/hash_index_utils.h"
+#include "storage/storage_structure/in_mem_page.h"
+#include "storage/storage_utils.h"
+#include "storage/wal/wal.h"
+#include "storage/wal/wal_record.h"
+#include "transaction/transaction.h"
+
+namespace kuzu {
+namespace storage {
+
+class OverflowFile;
+
+class OverflowFileHandle {
+
+public:
+    OverflowFileHandle(OverflowFile& overflowFile, PageCursor& nextPosToWriteTo)
+        : nextPosToWriteTo(nextPosToWriteTo), overflowFile{overflowFile} {}
+    // The OverflowFile stores the handles and returns pointers to them.
+    // Moving the handle would invalidate those pointers
+    OverflowFileHandle(OverflowFileHandle&& other) = delete;
+
+    std::string readString(transaction::TransactionType trxType, const common::ku_string_t& str);
+
+    bool equals(transaction::TransactionType trxType, std::string_view keyToLookup,
+        const common::ku_string_t& keyInEntry) const;
+
+    common::ku_string_t writeString(std::string_view rawString);
+    inline common::ku_string_t writeString(const char* rawString) {
+        return writeString(std::string_view(rawString));
+    }
+
+    void prepareCommit();
+    inline void checkpointInMemory() { pageCache.clear(); }
+    inline void rollbackInMemory(PageCursor nextPosToWriteTo) {
+        pageCache.clear();
+        this->nextPosToWriteTo = nextPosToWriteTo;
+    }
+
+private:
+    uint8_t* addANewPage();
+    void setStringOverflow(
+        const char* inMemSrcStr, uint64_t len, common::ku_string_t& diskDstString);
+
+    inline void read(transaction::TransactionType trxType, common::page_idx_t pageIdx,
+        const std::function<void(uint8_t*)>& func) const;
+
+private:
+    static const common::page_idx_t END_OF_PAGE =
+        common::BufferPoolConstants::PAGE_4KB_SIZE - sizeof(common::page_idx_t);
+    // This is the index of the last free byte to which we can write.
+    PageCursor& nextPosToWriteTo;
+    OverflowFile& overflowFile;
+    // Cached pages which have been written in the current transaction
+    std::unordered_map<common::page_idx_t, std::unique_ptr<InMemPage>> pageCache;
```

Comment:
You can rename `pageCache` to `pageWriteCache` to be more explicit that the cache is only triggered by page writes, but not reads.